### ISSUE TITLE
lora微调检测模型bug

### ISSUE THREAD

**Issue by AIFFFENG** [QwenLM/Qwen2-VL#190]
#121 为了进一步阐述我的问题，我在121的issus上添加了一个实验，我选取了4张图片作为训练集，训练了50个epoch，后续损失为0，然后合并lora权重，使用训练集的图片推理，按理说推理的结果应该完全一致，但是现在出现了两种情况，1.完全与标签一致。2.与标签差别巨大。
为了方面你们复现这个实验，目前的llama factory(0.9.1)与qwen2vl均为最新版本。训练采取你们提供的bash脚本，然后利用llama factory进行微调，
          --stage sft \
          --do_train \
          --model_name_or_path ... \
          --dataset test_finetune \
          --template qwen2_vl \
          --finetuning_type lora \
          --lora_target all \
          --lora_rank 64 \
          --lora_alpha 128 \
          --freeze_vision_tower false \
          --output_dir ... \
          --overwrite_cache \
          --overwrite_output_dir \
          --warmup_steps 100 \
          --weight_decay 0.1 \
          --preprocessing_num_workers 2 \
          --per_device_train_batch_size 2 \
          --gradient_accumulation_steps 1 \
          --ddp_timeout 90000 \
          --learning_rate 4e-5 \
          --lr_scheduler_type cosine \
          --logging_steps 1 \
          --cutoff_len 4096 \
          --save_steps 100 \
          --save_total_limit 100 \
          --plot_loss \
          --num_train_epochs 50 \
          --bf16
之后采取llamafactory官方脚本合并
      ### Note: DO NOT use quantized model or quantization_bit when merging lora adapters
      
      ### model
      model_name_or_path: ...
      adapter_name_or_path: ...
      template: qwen2_vl
      finetuning_type: lora
      
      ### export
      export_dir: ...
      export_size: 2
      export_device: cpu
      export_legacy_format: false

然后采取官方推理脚本推理
                from qwen_vl_utils import process_vision_info
                from transformers import AutoProcessor, AutoTokenizer, Qwen2VLForConditionalGeneration
                import os
                import shutil
                os.environ["CUDA_VISIBLE_DEVICES"]="5"
                model_path = ".."
                src_json = ".."
                model_json=os.path.join(model_path,"chat_template.json")
                shutil.copy(src_json,model_json)
                tokenizer = AutoTokenizer.from_pretrained(model_path)
                processor = AutoProcessor.from_pretrained(model_path)
                model = Qwen2VLForConditionalGeneration.from_pretrained(
                    model_path,
                    torch_dtype="auto",
                    #attn_implementation="flash_attention_2",
                    device_map="auto",
                )
                
                messages = [
                    [
                        {
                            "role": "user",
                            "content": [
                                {
                                    "type": "image",
                                    "image": ,
                                },
                                {
                                    "type": "text",
                                    "text": "...",
                                },
                            ],
                        },
                    ],
                ]
                
                # Preparation for inference
                text = processor.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
                image_inputs, video_inputs = process_vision_info(messages)
                inputs = processor(text=text, images=image_inputs, videos=video_inputs, padding=True, return_tensors="pt")
                inputs = inputs.to("cuda")
                
                # Inference: Generation of the output
                generated_ids = model.generate(**inputs, max_new_tokens=128)
                generated_ids_trimmed = [out_ids[len(in_ids) :] for in_ids, out_ids in zip(inputs.input_ids, generated_ids)]
                
                # Output
                output_text = tokenizer.decode(generated_ids_trimmed[0], skip_special_tokens=False, clean_up_tokenization_spaces=False)
                output_tokens = tokenizer.convert_ids_to_tokens(generated_ids_trimmed[0])
                print(output_text)
                print(output_tokens)
整个训练代码中唯一修改的点，是在     LLaMA-Factory/src/llamafactory/data/mm_plugin.py   中加了resize((448,448)）
results = []
        for image in images:
            if isinstance(image, str):
                image = Image.open(image)
            elif isinstance(image, dict):
                if image["bytes"] is not None:
                    image = Image.open(BytesIO(image["bytes"]))
                else:
                    image = Image.open(image["path"])

            if not isinstance(image, ImageObject):
                raise ValueError("Expect input is a list of Images, but got {}.".format(type(image)))
            image=image.resize((448,448))
            print(image.height,image.width)
            results.append(self._preprocess_image(image, **kwargs))
推理图片与训练图片一模一样，推理的图尺寸也为448*448。对于数据集的处理方式与qwenvl一致，只是更换了对应的特殊token与qwen2vl对齐

**Comment by AIFFFENG** [comment]
为了探究哪部分造成的bug，测试了Qwen2VL2B模型在数据集上的全量微调，测试结果没问题，怀疑是llama factory合并lora的问题

**Comment by ShuaiBai623** [comment]
https://github.com/huggingface/transformers/pull/33487/files 可以参考这个修改下

**Comment by AIFFFENG** [comment]
> https://github.com/huggingface/transformers/pull/33487/files 可以参考这个修改下

感谢回复，那么我lora模型是只需要更改推理代码，还是需要重新训练(因为我发现全量微调是没有问题的)?

**Comment by AIFFFENG** [comment]
> https://github.com/huggingface/transformers/pull/33487/files 可以参考这个修改下

感谢回复，修正后在小批量(4个数组)上实验，框的大小、位置正常！

**Comment by AIFFFENG** [comment]
> @AIFFFENG 请问是在训练和推理时都需要修正这部分代码吗

是的，需要重新训练和推理

**Comment by stomachacheGE** [comment]
> > https://github.com/huggingface/transformers/pull/33487/files 可以参考这个修改下
> 
> 感谢回复，修正后在小批量(4个数据)上实验，框的大小、位置正常！

这个bug只是在检测任务上才影响吗？其他qa任务需要处理不？

**Comment by AIFFFENG** [comment]
> > > https://github.com/huggingface/transformers/pull/33487/files 可以参考这个修改下
> > 
> > 
> > 感谢回复，修正后在小批量(4个数据)上实验，框的大小、位置正常！
> 
> 这个bug只是在检测任务上才影响吗？其他qa任务需要处理不？

目前在caption任务上测过，影响不大。感觉也是看qa任务需不需要精确的位置信息，毕竟检测任务也是组织成qa格式

**Comment by WangRongsheng** [comment]
@AIFFFENG 你好，是否可以出个详细的修改介绍，可以给该测试的example数据吗？

**Comment by thusinh1969** [comment]
Same here, very suddent, used to work Ok last few weeks. Any idea why bros ?

Steve

**Comment by zengxingchen** [comment]
你好，我参照你的方法训练至loss为0，再测试了一下训练集中的数据。发现输出的bbox还是不一致，已经把位置编码的部分加上了。你还有其他方向debug的建议吗。

**Comment by AIFFFENG** [comment]
> 你好，我参照你的方法训练至loss为0，再测试了一下训练集中的数据。发现输出的bbox还是不一致，已经把位置编码的部分加上了。你还有其他方向debug的建议吗。

是加上位置编码后重新训练了吗？

**Comment by AIFFFENG** [comment]
> 你好，我参照你的方法训练至loss为0，再测试了一下训练集中的数据。发现输出的bbox还是不一致，已经把位置编码的部分加上了。你还有其他方向debug的建议吗。

是加上位置编码后重新训练了吗？

**Comment by AIFFFENG** [comment]
> 你好，我参照你的方法训练至loss为0，再测试了一下训练集中的数据。发现输出的bbox还是不一致，已经把位置编码的部分加上了。你还有其他方向debug的建议吗。

是加上位置编码后重新训练还出现这个问题吗？

**Comment by zengxingchen** [comment]
> > 你好，我参照你的方法训练至loss为0，再测试了一下训练集中的数据。发现输出的bbox还是不一致，已经把位置编码的部分加上了。你还有其他方向debug的建议吗。
> 
> 是加上位置编码后重新训练还出现这个问题吗？

谢谢回复。是的。
我也测试了最新的transformer库中的`modeling_qwen2_vl.py`.
他们位置编码已经进一步修改为
```python
        # if we get 4D attention mask we cannot calculate rope deltas anymore. TODO @raushan fixme
        if position_ids is None and input_ids is not None and (attention_mask is None or attention_mask.ndim == 2):
            # calculate RoPE index once per generation in the pre-fill stage only
            if (cache_position is not None and cache_position[0] == 0) or self.rope_deltas is None:
                position_ids, rope_deltas = self.get_rope_index(
                    input_ids, image_grid_thw, video_grid_thw, attention_mask
                )
                self.rope_deltas = rope_deltas
            # then use the prev pre-calculated rope-deltas to get the correct position ids
            else:
                batch_size, seq_length, _ = inputs_embeds.shape
                delta = cache_position[0] + self.rope_deltas if cache_position is not None else 0
                position_ids = torch.arange(seq_length, device=inputs_embeds.device)
                position_ids = position_ids.view(1, -1).expand(batch_size, -1)
                if cache_position is not None:  # otherwise `deltas` is an int `0`
                    delta = delta.repeat_interleave(batch_size // delta.shape[0], dim=0)
                position_ids = position_ids.add(delta)
                position_ids = position_ids.unsqueeze(0).expand(3, -1, -1)
```
我测试了最新的上述代码，以及 https://github.com/huggingface/transformers/pull/33487/files 的版本，都还是有不一致。

**Comment by AIFFFENG** [comment]
> > > 你好，我参照你的方法训练至loss为0，再测试了一下训练集中的数据。发现输出的bbox还是不一致，已经把位置编码的部分加上了。你还有其他方向debug的建议吗。
> > 
> > 
> > 是加上位置编码后重新训练还出现这个问题吗？
> 
> 谢谢回复。是的。 我也测试了最新的transformer库中的`modeling_qwen2_vl.py`. 他们位置编码已经进一步修改为
> 
> ```python
>         # if we get 4D attention mask we cannot calculate rope deltas anymore. TODO @raushan fixme
>         if position_ids is None and input_ids is not None and (attention_mask is None or attention_mask.ndim == 2):
>             # calculate RoPE index once per generation in the pre-fill stage only
>             if (cache_position is not None and cache_position[0] == 0) or self.rope_deltas is None:
>                 position_ids, rope_deltas = self.get_rope_index(
>                     input_ids, image_grid_thw, video_grid_thw, attention_mask
>                 )
>                 self.rope_deltas = rope_deltas
>             # then use the prev pre-calculated rope-deltas to get the correct position ids
>             else:
>                 batch_size, seq_length, _ = inputs_embeds.shape
>                 delta = cache_position[0] + self.rope_deltas if cache_position is not None else 0
>                 position_ids = torch.arange(seq_length, device=inputs_embeds.device)
>                 position_ids = position_ids.view(1, -1).expand(batch_size, -1)
>                 if cache_position is not None:  # otherwise `deltas` is an int `0`
>                     delta = delta.repeat_interleave(batch_size // delta.shape[0], dim=0)
>                 position_ids = position_ids.add(delta)
>                 position_ids = position_ids.unsqueeze(0).expand(3, -1, -1)
> ```
> 
> 我测试了最新的上述代码，以及 https://github.com/huggingface/transformers/pull/33487/files 的版本，都还是有不一致。

那很奇怪呀，我当时加上位置编码后就变的正常了，我这边暂时没有其他解这个bug的思路，或许可以提个issue问问

**Comment by zengxingchen** [comment]
> > > > 你好，我参照你的方法训练至loss为0，再测试了一下训练集中的数据。发现输出的bbox还是不一致，已经把位置编码的部分加上了。你还有其他方向debug的建议吗。
> > > 
> > > 
> > > 是加上位置编码后重新训练还出现这个问题吗？
> > 
> > 
> > 谢谢回复。是的。 我也测试了最新的transformer库中的`modeling_qwen2_vl.py`. 他们位置编码已经进一步修改为
> > ```python
> >         # if we get 4D attention mask we cannot calculate rope deltas anymore. TODO @raushan fixme
> >         if position_ids is None and input_ids is not None and (attention_mask is None or attention_mask.ndim == 2):
> >             # calculate RoPE index once per generation in the pre-fill stage only
> >             if (cache_position is not None and cache_position[0] == 0) or self.rope_deltas is None:
> >                 position_ids, rope_deltas = self.get_rope_index(
> >                     input_ids, image_grid_thw, video_grid_thw, attention_mask
> >                 )
> >                 self.rope_deltas = rope_deltas
> >             # then use the prev pre-calculated rope-deltas to get the correct position ids
> >             else:
> >                 batch_size, seq_length, _ = inputs_embeds.shape
> >                 delta = cache_position[0] + self.rope_deltas if cache_position is not None else 0
> >                 position_ids = torch.arange(seq_length, device=inputs_embeds.device)
> >                 position_ids = position_ids.view(1, -1).expand(batch_size, -1)
> >                 if cache_position is not None:  # otherwise `deltas` is an int `0`
> >                     delta = delta.repeat_interleave(batch_size // delta.shape[0], dim=0)
> >                 position_ids = position_ids.add(delta)
> >                 position_ids = position_ids.unsqueeze(0).expand(3, -1, -1)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 我测试了最新的上述代码，以及 https://github.com/huggingface/transformers/pull/33487/files 的版本，都还是有不一致。
> 
> 那很奇怪呀，我当时加上位置编码后就变的正常了，我这边暂时没有其他解这个bug的思路，或许可以提个issue问问
好的，谢谢回复。
